src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBindingHelper.cs

```diff
@@ -659,27 +659,76 @@ internal static string GetPropertyName(Expression expression)
+            }
+
+            if (modelstate == null)
+            {
+                throw new ArgumentNullException(nameof(modelstate));
+            }
+
             if (string.IsNullOrEmpty(modelKey))
             {
-                var modelMetadata = metadataProvider.GetMetadataForType(modelType);
-                var elementMetadata = modelMetadata.ElementMetadata;
-                if (elementMetadata != null)
+                // If model key is empty, we have to do a best guess to try and clear the appropriate
+                // keys. Clearing the empty prefix would clear the state of ALL entries, which might wipe out
+                // data from other models.
+                if (modelMetadata.IsEnumerableType)
                 {
-                    modelMetadata = elementMetadata;
+                    // We expect that any key beginning with '[' is an index. We can't just infer the indexes
+                    // used, so we clear all keys that look like <empty prefix -> index>.
+                    //
+                    // In the unlikely case that multiple top-level collections where bound to the empty prefix,
+                    // you're just out of luck.
+                    foreach (var kvp in modelstate)
+                    {
+                        if (kvp.Key.Length > 0 && kvp.Key[0] == '[')
+                        {
+                            // Starts with an indexer
+                            kvp.Value.Errors.Clear();
+                            kvp.Value.ValidationState = ModelValidationState.Unvalidated;
+                        }
+                    }
                 }
-
-                foreach (var property in modelMetadata.Properties)
+                else if (modelMetadata.IsComplexType)
                 {
-                    var childKey = property.BinderModelName ?? property.PropertyName;
-                    modelstate.ClearValidationState(childKey);
+                    foreach (var property in modelMetadata.Properties)
+                    {
+                        modelstate.ClearValidationState(property.BinderModelName ?? property.PropertyName);
+                    }
+                }
+                else
+                {
+                    // Simple types bind to a single entry. So clear the entry with the empty-key, in the
+                    // unlikely event that it has errors.
+                    var entry = modelstate[string.Empty];
+                    if (entry != null)
+                    {
+                        entry.Errors.Clear();
+                        entry.ValidationState = ModelValidationState.Unvalidated;
+                    }
                 }
             }
             else
             {
+                // If model key is non-empty, we just want to clear all keys with that prefix. We expect
+                // model binding to have only used this key (and suffixes) for all entries related to
+                // this model.
```

Comment:
Liking these comments. Now we might remember what we decided a few months from now -- and why :clap: